File: tests/test_frame.py
from patchyAnalysisTools import Frame, Patches, Clusters

class TestFrame:
    def test_init(self):
        coords = [[1, 1, 1], [2, 2, 2]]
        orients = [[0,0,0],[0,0,0]]
        frame = Frame(coords, orients, [3, 3, 3])

        assert frame.get_N() == 2
        assert frame.get_time_stamp() == 0
        assert len(frame.get_coordinates()) == 2
        assert len(frame.get_orientations_as_angles()) == 2

    def test_set_types(self):
        coords = [[1, 1, 1], [2, 2, 2]]
        orients = [[0,0,0],[0,0,0]]
        frame = Frame(coords, orients, [3, 3, 3])

        types = frame.get_types()
        assert types[0] == 0
        assert types[1] == 0

        types_2 = [1, 5]
        frame.set_types(types_2)
        assert frame.get_types()[0] == 1
        assert frame.get_types()[1] == 5

        frame.set_time_stamp(5)
        assert frame.get_time_stamp() == 5

    def test_interactions(self):
        coords = []
        orients = []
        sep = 1.05
        for i in range(3):
            for j in range(3):
                for k in range(3):
                    coords.append([i * sep, j * sep, k * sep])
                    orients.append([0,0,0])
        cell = [sep * 3, sep * 3, sep * 3]

        patches = Patches()
        patches.add_patch(1, 1.1, 0.92, [1,0,0], 0)
        patches.add_patch(1, 1.1, 0.92, [-1,0,0], 0)
        patches.add_patch(1, 1.1, 0.92, [0,1,0], 0)
        patches.add_patch(1, 1.1, 0.92, [0,-1,0], 0)
        patches.add_patch(1, 1.1, 0.92, [0,0,1], 0)
        patches.add_patch(1, 1.1, 0.92, [0,0,-1], 0)
        patches.make_all_patches_adjacent()

        frame = Frame(coords, orients, cell)
        frame.determine_bond_list(patches)
        bond_list = frame.get_bond_list()
        assert len(bond_list) == len(coords) * 3

    def test_clusters(self):
        coords = []
        orients = []
        sep = 1.05
        for i in range(3):
            coords.append([i * sep, 0, 0])
            orients.append([0,0,0])
        cell = [sep * 3, sep * 3, sep * 3]

        patches = Patches()
        patches.add_patch(1, 1.1, 0.92, [1,0,0], 0)
        patches.add_patch(1, 1.1, 0.92, [-1,0,0], 0)
        patches.make_all_patches_adjacent()

        frame = Frame(coords, orients, cell)
        frame.determine_bond_list(patches)
        frame.determine_clusters()
        frame.determine_percolation(patches)
        clusters = frame.get_clusters()
        clusters_list = clusters.get_clusters()
        assert len(clusters_list) == 1
        assert len(clusters_list[0]) == 3

File: tests/cpp/test_frame.cpp
#include "frame.hpp"
#include <gtest/gtest.h>
#include <math.h>
#include <iostream>

TEST(FrameTests, init){
    Eigen::Vector3d pos(1,1,1);
    Rotation rot(0,0,0);
    std::vector<Eigen::Vector3d> coords;
    std::vector<Rotation> orients;
    int N = 10;
    for(int i = 0; i < N; i++){
        coords.push_back(pos);
        orients.push_back(rot);
    }

    Eigen::Vector3d cell(2,2,2);

    Frame my_frame(coords, orients, cell);

    ASSERT_EQ(my_frame.get_N(), N);

    std::vector<double> cell_vec = my_frame.get_cell_as_list();
    for(int i = 0; i < 3; i++) ASSERT_EQ(cell_vec[i], cell(i));

    std::vector<int> types = my_frame.get_types();
    for( auto t : types){
        ASSERT_EQ(t, 0);
    }

    std::vector<std::vector<double> > angles = my_frame.get_orientations_as_list_of_angles();
    for( auto t : angles){
        for(int i = 0; i < 3; i++) ASSERT_EQ(t[i], 0);
    }
}

TEST(FrameTests, setters){
    Eigen::Vector3d pos(1,1,1);
    Rotation rot(0,0,0);
    std::vector<Eigen::Vector3d> coords;
    std::vector<Rotation> orients;
    int N = 10;
    for(int i = 0; i < N; i++){
        coords.push_back(pos);
        orients.push_back(rot);
    }

    Eigen::Vector3d cell(2,2,2);

    Frame my_frame(coords, orients, cell);

    ASSERT_EQ(my_frame.get_time_stamp(), 0);
    my_frame.set_time_stamp(128);
    ASSERT_EQ(my_frame.get_time_stamp(), 128);

    std::vector<int> types;
    for(int i = 0; i < N; i++){
        types.push_back(i);
    }
    my_frame.set_types(types);
    std::vector<int> types_g = my_frame.get_types();
    for(int i = 0; i < N; i++) ASSERT_EQ(types[i], types_g[i]);
}

TEST(FrameTests, test_interaction){
    std::vector<Eigen::Vector3d> coords;
    std::vector<Rotation> orients;
    Rotation rot(0,0,0);
    double sep = 1.05;
    for(int i = 0; i < 3; i++){
        for (int j = 0; j < 3; j++){
            for (int k = 0; k < 3; k++){
                Eigen::Vector3d pos(i * sep, j * sep, k * sep);
                coords.push_back(pos);
                orients.push_back(rot);
            }
        }
    }

    Eigen::Vector3d cell(3 * sep, 3 * sep, 3 * sep);
    Frame frame(coords, orients, cell);

    // create the patch obj
    Patches patches;
    Eigen::Vector3d p0(1,0,0);
    patches.add_patch(1, 1.1, 0.92, p0, 0);
    Eigen::Vector3d p1(-1,0,0);
    patches.add_patch(1, 1.1, 0.92, p1, 0);
    Eigen::Vector3d p2(0,1,0);
    patches.add_patch(1, 1.1, 0.92, p2, 0);
    Eigen::Vector3d p3(0,-1,0);
    patches.add_patch(1, 1.1, 0.92, p3, 0);
    Eigen::Vector3d p4(0,0,1);
    patches.add_patch(1, 1.1, 0.92, p4, 0);
    Eigen::Vector3d p5(0,0,-1);
    patches.add_patch(1, 1.1, 0.92, p5, 0);

    ASSERT_TRUE(frame.does_ij_interact(0, 1, patches));
    ASSERT_TRUE(frame.does_ij_interact(0, 3, patches));
    ASSERT_TRUE(frame.does_ij_interact(0, 2, patches));
    ASSERT_FALSE(frame.does_ij_interact(0, 4, patches));
    ASSERT_TRUE(frame.does_ij_interact(0, 9, patches));
}

TEST(FrameTests, test_bond_list){
    std::vector<Eigen::Vector3d> coords;
    std::vector<Rotation> orients;
    Rotation rot(0,0,0);
    double sep = 1.05;
    for(int i = 0; i < 3; i++){
        for (int j = 0; j < 3; j++){
            for (int k = 0; k < 3; k++){
                Eigen::Vector3d pos(i * sep, j * sep, k * sep);
                coords.push_back(pos);
                orients.push_back(rot);
            }
        }
    }

    Eigen::Vector3d cell(3 * sep, 3 * sep, 3 * sep);
    Frame frame(coords, orients, cell);

    // create the patch obj
    Patches patches;
    Eigen::Vector3d p0(1,0,0);
    patches.add_patch(1, 1.1, 0.92, p0, 0);
    Eigen::Vector3d p1(-1,0,0);
    patches.add_patch(1, 1.1, 0.92, p1, 0);
    Eigen::Vector3d p2(0,1,0);
    patches.add_patch(1, 1.1, 0.92, p2, 0);
    Eigen::Vector3d p3(0,-1,0);
    patches.add_patch(1, 1.1, 0.92, p3, 0);
    Eigen::Vector3d p4(0,0,1);
    patches.add_patch(1, 1.1, 0.92, p4, 0);
    Eigen::Vector3d p5(0,0,-1);
    patches.add_patch(1, 1.1, 0.92, p5, 0);

    // determine bond_list
    frame.determine_bond_list(patches);
    std::vector<std::vector<int> > bond_list = frame.get_bond_list();
    ASSERT_EQ(bond_list.size(), coords.size() * 3);
}

int main(int argc, char** argv){
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}